python/ray/tune/trial_runner.py

```diff
@@ -463,13 +466,15 @@ def step(self):
         # This will contain the next trial to start
         next_trial = self._get_next_trial()  # blocking
 
-        # Create pending trials
-        num_pending_trials = len(
-            [t for t in self._trials if t.status == Trial.PENDING])
-        while num_pending_trials < self._max_pending_trials:
-            if not self._update_trial_queue(blocking=False):
-                break
-            num_pending_trials += 1
+        # Create pending trials. If the queue was updated before, only
+        # continue updating if this was successful (next_trial is not None)
+        if not self._updated_queue or (self._updated_queue and next_trial):
```

Comment:
I think in the future we'll want to simplify the logic between `get_next_trial` and this code block